src/cryptonote_basic/cryptonote_format_utils.cpp

```diff
@@ -866,45 +889,112 @@ namespace cryptonote
     return res;
   }
   //---------------------------------------------------------------
-  bool is_out_to_acc(const account_keys& acc, const txout_to_key& out_key, const crypto::public_key& tx_pub_key, const std::vector<crypto::public_key>& additional_tx_pub_keys, size_t output_index)
+  void set_tx_out(const uint64_t amount, const crypto::public_key output_public_key, const bool use_view_tags, const crypto::view_tag view_tag, tx_out &out)
+  {
+    out.amount = amount;
+    if (use_view_tags)
+    {
+      txout_to_tagged_key ttk;
+      ttk.key = output_public_key;
+      ttk.view_tag = view_tag;
+      out.target = ttk;
+    }
+    else
+    {
+      txout_to_key tk;
+      tk.key = output_public_key;
+      out.target = tk;
+    }
+  }
+  //---------------------------------------------------------------
+  bool check_view_tags_valid(const transaction& tx, uint8_t hf_version)
+  {
+    for (const auto &o: tx.vout)
+    {
+      if (hf_version >= HF_VERSION_VIEW_TAGS)
+      {
+        // don't allow old outputs that don't have view tags after HF_VERSION_VIEW_TAGS
+        CHECK_AND_ASSERT_MES(o.target.type() != typeid(txout_to_key), false, "wrong variant type: "
+          << "txout_to_key, expected txout_to_tagged_key in transaction id=" << get_transaction_hash(tx));
+      }
+      else
+      {
+        // don't allow view tags until HF_VERSION_VIEW_TAGS
+        CHECK_AND_ASSERT_MES(o.target.type() != typeid(txout_to_tagged_key), false, "wrong variant type: "
+          << "txout_to_tagged_key not supported yet, expected txout_to_key in transaction id=" << get_transaction_hash(tx));
+      }
```

Comment:
```suggestion
        CHECK_AND_ASSERT_MES(o.target.type() == typeid(txout_to_key), false, "wrong variant type... "
          << "only txout_to_key is supported in transaction id=" << get_transaction_hash(tx));
      }
```

Pre-fork should only care about pre-fork things.

Reply:
This check didn't exist before, so I didn't add it. There are still the other 2 types `txout_to_script` and `txout_to_scripthash` that this would prevent at this stage, although those types [can't even be added to lmdb](https://github.com/monero-project/monero/blob/6d94953be259c6bf4a1407ec552a570a66b6179d/src/blockchain_db/lmdb/db_lmdb.cpp#L1048-L1049) -- so maybe makes sense to add this check on these grounds?

Reply:
This logic branch doesn't know about `txout_to_tagged_key`, since `txout_to_tagged_key` doesn't show up until `HF_VERSION_VIEW_TAGS`.

Reply:
Sorry wasn't very clear what I meant. 2 of the 3 callers to this function are not near to any other functions that require the `txout_to_key` type (e.g. currently `blockchain.cpp::check_tx_outputs` doesn't require outputs have the `txout_to_key` type). So I didn't want to add logic requiring `txout_to_key` in this code path because of that. But like mentioned [here](https://github.com/monero-project/monero/pull/8061#discussion_r749706312), maybe this is ok to do. I'll do more digging on it

Reply:
Any logic branch that depends on a hardfork is in that hardfork's 'domain', regardless of context. Since `txout_to_key` is the only type that should exist in hardforks pre-`HF_VERSION_VIEW_TAGS`, that is what should be enforced. Redundant type checks are fine.

Reply:
My only concern was if `txout_to_script` and `txout_to_script_hash` have or can be used for something, and should make it past this logic branch (and perhaps fail elsewhere or something). That's what I was going to dig into to. I assume not, but didn't want to take a chance if that assumption was incorrect

Reply:
Ah gotcha

Reply:
Done